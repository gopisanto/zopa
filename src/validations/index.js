export const REQUIRED_FIELD = 'This field is required.';
export const ONLY_ALPHABETS = 'Only alphabets, no numbers or special characters allowed.';
export const INVALID_EMAIL = 'Invalid Email ID.';
export const AMOUNT_REQUIRED = 'Please enter amount to transfer';
export const POSITIVE_AMOUNT_REQUIRED = 'Enter proper amount to transfer';
export const INSUFFICIENT_FUNDS = 'Insufficient funds.';

export const validateName = name => /^[a-zA-Z ]+$/.test(name);

export const validateEmail = email => /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(email);

export const validate = (values, { totalLeft }) => {
  const errors = {};
  const { name, email, amount } = values;

  if (!name) {
    errors.name = REQUIRED_FIELD;
  } else if (!validateName(name)) {
    errors.name = ONLY_ALPHABETS;
  }

  if (!email) {
    errors.email = REQUIRED_FIELD;
  } else if (!validateEmail(email)) {
    errors.email = INVALID_EMAIL;
  }

  if (amount !== 0 && !amount) {
    errors.amount = REQUIRED_FIELD;
  } else if (!Number.isFinite(Number(amount))) {
    errors.amount = POSITIVE_AMOUNT_REQUIRED;
  } else if (amount === 0) {
    errors.amount = AMOUNT_REQUIRED;
  } else if (amount < 0) {
    errors.amount = POSITIVE_AMOUNT_REQUIRED;
  } else if (totalLeft - amount < 0) {
    errors.amount = INSUFFICIENT_FUNDS;
  }

  return errors;
};
