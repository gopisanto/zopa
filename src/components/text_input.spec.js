import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import { cloneDeep } from 'lodash';

import TextInput from './text_input';
import { REQUIRED_FIELD, ONLY_ALPHABETS } from '../validations';

Enzyme.configure({ adapter: new Adapter() });

const noop = () => null;

describe('TextInput component', () => {
  test('should render without error', () => {
    const field = {
      meta: {
        touched: false,
        error: undefined,
        active: false,
      },
      name: 'name',
      label: 'Name',
      type: 'text',
      input: {value: "ZOPA", name: "name", onChange: noop}
    };
    const component = renderer.create(
      <TextInput field={field} />
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  test('should render with matching error messages based on text input value', () => {
    const fieldWithEmptyValue = {
      meta: {
        touched: true,
        error: REQUIRED_FIELD,
        active: false,
      },
      name: 'name',
      label: 'Name',
      type: 'text',
      input: {value: "", name: "name", onChange: noop},
    };
    const component = mount(<TextInput field={fieldWithEmptyValue} />);
    let fieldWithInvalidChars = cloneDeep(fieldWithEmptyValue);
    fieldWithInvalidChars.meta.error = ONLY_ALPHABETS;
    fieldWithInvalidChars.input.value = '@santosh-kumar';
    const component1 = mount(<TextInput field={fieldWithInvalidChars} />);

    expect(component.find('span').text()).toBe(REQUIRED_FIELD);
    expect(component1.find('span').text()).toBe(ONLY_ALPHABETS);
  });

  test('should render with error message', () => {
    const field = {
      meta: {
        touched: true,
        error: REQUIRED_FIELD,
        active: false,
      },
      name: 'name',
      label: 'Name',
      type: 'text',
      input: {value: "", name: "name", onChange: noop},
    };
    const component = renderer.create(
      <TextInput field={field} />
    ).toJSON();

    expect(component).toMatchSnapshot();
  });
});
