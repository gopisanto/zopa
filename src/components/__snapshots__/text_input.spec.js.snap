// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput component should render with error message 1`] = `
<div
  className="ti"
>
  <label
    className="ti-lbl"
    htmlFor="name"
  >
    Name
  </label>
  <div>
    <input
      autoComplete="none"
      className="ti-field err"
      name="name"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <span
    className="err-msg"
  >
    This field is required.
  </span>
</div>
`;

exports[`TextInput component should render without error 1`] = `
<div
  className="ti"
>
  <label
    className="ti-lbl"
    htmlFor="name"
  >
    Name
  </label>
  <div>
    <input
      autoComplete="none"
      className="ti-field"
      name="name"
      onChange={[Function]}
      type="text"
      value="ZOPA"
    />
  </div>
</div>
`;
